### ISSUE TITLE
ability to query from an array of maps/proplists/etc

### ISSUE THREAD

**Issue by juantascon** [nalundgaard/jsn#10]
Having an interface to query lists of maps/proplists/etc would be very handy, jwalk for instance implements something like that.

Example:
```
Objs = [
[{<<"key1">>, <<"value1">>}],
[{<<"key1">>, <<"value2">>}],
[{<<"key1">>, <<"value3">>}]
].
jwalk:get({<<"key1">>}, Objs).
[<<"value1">>,<<"value2">>,<<"value3">>]
```
The function doesn't have to be `jsn:get()` but I guess something like `jsn:get_from_list()` or `jsn:select()` could do it too.

I'd lile to submit a pullrq if you guys approve an interface that aligns with the roadmap
Thank you

**Comment by jadeallenx** [comment]
Like the name `select` as a function to implement this. Sounds handy to me too.

**Comment by nalundgaard** [comment]
Yes, I like the name too. @juantascon and I were talking about this earlier.

I'm not too interested in the "base" case, e.g.

```erlang
select(Path, Elements) ->
    select(Path, Elements, undefined).

select(Path, Elements, Default) ->
   [get(Path, V) || V <- Elements].
```

IMO that's basically best as a one-liner as-needed.

I would be interested in building a `select` function that could do the above and be compatible with future expansion. e.g.:

* Taking flexible output transformation:
  * `identity` (return full value unmodified),
  * `Path` (return the value from the path—with an optional `Default` in the input)
  * `Paths` (return an array of extracted values for each element)
* A higher arity that also takes a list of filters/criteria. The `select` function would always apply the output "transformations", but the actual values would be filtered out such that only objects meeting the given criteria would be met. All criteria would take a `Path` as a first element, and then a condition to go with it—I assume equivalence would be the main one, but it might be good to leave the syntax of the criteria open to more sophisticated things.

**Comment by nalundgaard** [comment]
@mrallen1 @juantascon what to you think about something along these lines?

```erlang
-type select_output() :: identity | {path, path()} | {paths, paths()}.
-type select_numeric_comparison_op() :: eq | ne | le | lt | ge | gt.
-type select_string_comparison_op() :: contains | not_contains | begins_with | ends_with.
-type select_condition() :: {path(), term()} |
                            {path(), select_numeric_comparison_op(), json_number()} |
                            {path(), select_string_comparison_op(), json_string()} |
                            {path(), between, {json_number(), json_number()}} |
                            {path(), in, [json_term()]} |
                            {and, select_conditions()} |
                            {or, select_conditions()}.
-type select_conditions() :: [select_condition()].

-spec select(select_output(), json_array()) -> json_array().
select(OutputSpec, Elements) ->
    select(OutputSpec, [], Elements).

-spec select(select_output(), select_conditions(), json_array()) -> json_array().
...
```

**Comment by nalundgaard** [comment]
I should clarify, I think this might be a worthwhile spec to work towards—not necessarily needed for an initial version that could simply implement `select/2` in a forwards-compatible way. Even `select/3` could be implemented in a piecemeal way. For example, we could start with just `select_condition() :: {path(), json_term()}`.

**Comment by nalundgaard** [comment]
I think I messed up `select_output()` a bit. Needs to also support defaults. That seems obvious enough for the singular case, maybe not so much for the `paths` case.

Also, `path | paths` is probably not the right language. `value | values` looks more right to me.

```erlang
-type select_output() :: identity | {value, path()} | {value, path(), Default :: term()} | 
                         {values, [path() | {default, path(), Default :: term()}]}.
```

**Comment by jadeallenx** [comment]
I would strongly prefer that predicates and transformations are specified as closures.

```erlang
Paths = #{
  <<"foo.bar">> => fun(V) when is_integer(V) andalso V < 100 -> V*V
                      (V) when is_integer(V) -> V end,
  <<"baz.qux">> => fun(V) when is_binary(V) ->  <<"appending ", V/binary>> end,
  <<"hoge">>    => fun(V) -> V end,
},
Transformed = jsn:select(Json, Paths)
``` 

Just spitballing and I'm kind of preoccupied with a screenhero session so maybe I'm just full of crap.

**Comment by juantascon** [comment]
lgtm

**Comment by nalundgaard** [comment]
@mrallen1 I'm not sure. If I understand you right, it sure looks like you're looking to do transformations, not selection. Can you give me an example of how you would express @juantascon's initial ask in this manner?

**Comment by jadeallenx** [comment]
I am probably confusing things.

I was just thinking if I was going to select a value from a path across a list of objects and transform it, I would rather specify the paths and the transformations using a map where the keys are the path and the value is a closure which implements a transformation or a predicate to apply to the value at that path (if it exists).

It occurs to me that maybe the value is itself a map with keys like `predicate`, `transformation` or `default` or something like that.

```erlang
Paths = #{
  <<"foo">> => #{ default => baz },
  <<"bar">> => #{ predicate => fun(V) when is_integer(V) -> true; (V) -> false end,
                  default => 0 }
}
jsn:select(Paths, ObjectList)
```

And the behavior is an output set where each object either has the value of the key `foo` or `baz` and for the key `bar`, the output has the value of `bar` if its an integer, otherwise it has 0.

Is that what you're getting at with predicates and transformations with selection? 

Maybe I just fundamentally misunderstand the feature.

**Comment by nalundgaard** [comment]
I guess that makes sense, but I guess my question would be—if you do that—what about the "base case" of wanting to select one thing. As in. I have a list of objects, and I want to return the symmetric list of values for a single key:

For example, take this list of example objects:

```erlang
Objects = [#{<<"key1">> => <<"value11">>,
             <<"key2">> => <<"value21">>},
           #{<<"key1">> => <<"value12">>,
             <<"key2">> => <<"value22">>},
           #{<<"key1">> => <<"value13">>,
             <<"key2">> => <<"value23">>}].
```

I think @juantascon has the use case of wanting a function that can do things like...

```erlang
%% extract a single key from a list of objects: <<"key1">>
Expected1 = [<<"value11">>,<<"value12">>,<<"value13">>].
Expected1 = jsn:select(ThingContainingKey1, Objects).
```

I'm less sure of this secondary use case, but I think this would be another thing desired:

```erlang
%% extract a list of keys from a list of objects: [<<"key1">>, <<"key2">>]
Expected2 = [[<<"value11">>, <<"value21">>],
             [<<"value12">>, <<"value22">>],
             [<<"value13">>, <<"value23">>]].
Expected2 = jsn:select(ThingContainingKey1AndKey2, Objects).
```

Looking at your suggestion of taking a map of keys and transformations/predicates/defaults, I don't understand how you would map whole objects into a single value: it looks like what you're suggesting would just take a list of objects and transform each one using the given criteria. I think we already have that, and I think we already have that—see `jsn:transform/2`. I think we could improve THAT function by making it support maps rather than proplists to specify transformations, but that seems like a really different use case than what this is talking about.

**Comment by jadeallenx** [comment]
OK, fair enough.

I don't understand what you meant by predicates then with the `ne` `ge` stuff above... can you help me understand that better?

**Comment by nalundgaard** [comment]
Sure. Use cases like this:

```erlang
Objects = [#{<<"username">> => <<"Alice">>,
             <<"created">> => #{<<"by">> => <<"Dana">>,
                                <<"at">> => 1503702360}},
           #{<<"username">> => <<"Bob">>,
             <<"created">> => #{<<"by">> => <<"Eve">>,
                                <<"at">> => 1503702417}},
           #{<<"username">> => <<"Carl">>,
             <<"created">> => #{<<"by">> => <<"Frank">>,
                                <<"at">> => 1503703417}}].

[<<"Alice">>, <<"Bob">>, <<"Carl">>] = jsn:select({path, <<"username">>}, Objects).

[<<"Alice">>] = jsn:select({path, <<"username">>},
                           [{<<"created.at">>, lt, 1503702416}],
                           Objects).

[[<<"Bob">>, <<"Eve">>]] = jsn:select({paths, [<<"username">>,
                                               {<<"created">>, <<"by">>}]},
                                      [{<<"created.at">>, 1503702417}]).
```

The basic idea is that you can "select" a single field or list of fields. Maybe a transformation could be defined there as well, optionally.

Then, is seems to me, that filters/conditions would be applied per-object (rather, per array element). Of course, I'm not attached to making a whole big grammar of SQLesque queries—we could also just handle the simple case (tuple of path and exact match, like `{<<"created.at">>, 1503702417}`), and then let the user pass their own funs e.g. tuple of path and fun, like `{<<"created.at">>, fun(V) -> V < 1503702416 end}` or a whole-object fun, like `fun(T) -> jsn:get(<<"created.at">>, T) < 1503702416 end`. The idea here is that there is a way to pass conditions for whether certain elements would be included in the results at all—not whether they're transformed, or some value is excluded/set to a default.

**Comment by jadeallenx** [comment]
OK. I'd rather pass a closure than build some kind of faux grammar.

**Comment by nalundgaard** [comment]
Fair enough. What about this?

```erlang
-type value_selection() :: {value, path()} | {value, path(), Default :: term()}.
-type selection() :: identity | value_selection() | [value_selection()].
-type condition() :: {path(), Value :: json_term()} | %% exact match syntactic sugar
                     {path(), function((Value :: json_term()) -> boolean())} | %% boolean function with input of the value at the path
                     function(Element :: json_term()) -> boolean(). %% boolean function that takes a whole array element
-type conditions() :: [condition()].


-spec select(selection(), json_array()) -> json_array().
select(OutputSpec, Elements) ->
    select(OutputSpec, [], Elements).

-spec select(selection(), conditions(), json_array()) -> json_array().
...
```

Some usage examples:

```erlang
Objects = [#{<<"username">> => <<"Alice">>,
             <<"created">> => #{<<"by">> => <<"Dana">>,
                                <<"at">> => 1503702360}},
           #{<<"username">> => <<"Bob">>,
             <<"created">> => #{<<"by">> => <<"Eve">>,
                                <<"at">> => 1503702417}},
           #{<<"username">> => <<"Carl">>,
             <<"created">> => #{<<"by">> => <<"Frank">>,
                                <<"at">> => 1503703417}},
           #{<<"created">> => #{<<"by">> => <<"Gretchen">>,
                                <<"at">> => 1503704439}}].

[<<"Alice">>, <<"Bob">>, <<"Carl">>] = jsn:select({value, <<"username">>}, Objects).

[<<"Alice">>] = jsn:select({value, <<"username">>},
                           [{<<"created.at">>, fun(V) -> V < 1503702416 end}
                            fun(E) -> jsn:get(<<"created.at">>, E) == <<"Dana">> end],
                           Objects).

[[<<"Bob">>, <<"Eve">>]] = jsn:select([{value, <<"username">>, <<"No Name">>},
                                       {value, {<<"created">>, <<"by">>}}],
                                      [{<<"created.at">>, 1503702417}]).

[#{<<"created">> => #{<<"by">> => <<"Gretchen">>,
                      <<"at">> => 1503703417}}] = jsn:select(identity,
                                                             [{<<"username">>, undefined}],
                                                             Objects).

[<<"No Name">>] = jsn:select({value, <<"username">>, <<"No Name">>},
                             [{<<"username">>, undefined}],
                             Objects).
```

**Comment by jadeallenx** [comment]
Sounds good

**Comment by nalundgaard** [comment]
@juantascon I think the above is a good start. What do you think?

If you like it, are you interested in implementing it, or should I? Let me know if you prefer me to do it, and I can take a stab at it.

**Comment by juantascon** [comment]
lgtm and yes I'd like to give it a try

**Comment by juantascon** [comment]
initial implementation: https://github.com/nalundgaard/jsn/pull/13

**Comment by nalundgaard** [comment]
Resolved by #13
